nupic/embodied/models/multiheaded_dendrite_mlp.py

```diff
@@ -0,0 +1,125 @@
+
+from garage.torch import NonLinearity
+from nupic.research.frameworks.dendrites import (
+    AbsoluteMaxGatingDendriticLayer,
+    DendriticAbsoluteMaxGate1d,
+    DendriticGate1d,
+)
+from nupic.torch.modules import KWinners
+
+
+class MultiHeadedDendriticMLP(nn.Module):
+    """
+    A dendritic network which is similar to a MLP with a two hidden layers, except that
+    activations are modified by dendrites. The context input to the network is used as
+    input to the dendritic weights.
+                    _____
+                   |_____|    # Classifier layer, no dendrite input
+                      ^
+                      |
+                  _________
+    context -->  |_________|  # Second linear layer with dendrites
+                      ^
+                      |
+                  _________
+    context -->  |_________|  # First linear layer with dendrites
+                      ^
+                      |
+                    input
+    """
+
+    def __init__(self,
+                 input_size,
+                 num_heads,
+                 output_dims,
+                 dim_context,
+                 hidden_sizes=(32, 32),
+                 num_segments=(5, 5),
+                 sparsity=0.5,
+                 kw=False,
+                 relu=False,
+                 output_nonlinearities=None,
+                 dendritic_layer_class=AbsoluteMaxGatingDendriticLayer):
+
+        assert dendritic_layer_class in {AbsoluteMaxGatingDendriticLayer,
+                                         DendriticAbsoluteMaxGate1d,
+                                         DendriticGate1d}
+
+        # The nonlinearity can either be k-Winners or ReLU, but not both
+        assert not (kw and relu)
+        assert num_heads == len(output_dims)
+
+        super().__init__()
+
+        self.num_heads = num_heads
+        self.num_segments = num_segments
+        self.input_size = input_size
+        self.hidden_sizes = hidden_sizes
+        self.output_dims = output_dims
+        self.dim_context = dim_context
+        self.kw = kw
+        self.relu = relu
+
+        self._layers = []
+        self._activations = []
+        prev_dim = input_size
+        for i in range(len(hidden_sizes)):
+            curr_dend = dendritic_layer_class(
+                module=nn.Linear(prev_dim, hidden_sizes[i]),
+                num_segments=num_segments[i],
+                dim_context=dim_context,
+                module_sparsity=sparsity,
+                dendrite_sparsity=sparsity
+            )
+            if kw:
+                curr_activation = KWinners(n=hidden_sizes[i],
+                                           percent_on=0.05,
+                                           k_inference_factor=1.0,
+                                           boost_strength=0.0,
+                                           boost_strength_factor=0.0)
+            else:
+                curr_activation = nn.ReLU()
+
+            self._layers.append(curr_dend)
```

Comment:
This is a little tricky (as we found out on Friday) since Dendrite takes in both feedforward input and context, so this is a little cleaner for now I think.